src/apis/impl/DataOwnerApiImpl.ts

```diff
@@ -65,30 +67,68 @@ export class DataOwnerApiImpl implements DataOwnerApi {
     return dataOwnerId
   }
 
+  private _getDataOwnerPublicKeys(dataOwner: Patient | Device | HealthcareParty): string[] {
+    return [...Object.keys(dataOwner.aesExchangeKeys ?? {}), dataOwner.publicKey ?? ''].filter((k) => k !== '')
+  }
+
   async initCryptoFor(
     user: User,
     overwriteExistingKeys: boolean = false,
     keyPair?: { publicKey: string; privateKey: string }
   ): Promise<{ publicKey: string; privateKey: string }> {
     const dataOwnerId = this.getDataOwnerIdOf(user)
-
-    const { publicKey, privateKey } = keyPair ?? (await this._generateKeyPair())
-    const jwk = {
-      publicKey: spkiToJwk(hex2ua(publicKey)),
-      privateKey: pkcs8ToJwk(hex2ua(privateKey)),
-    }
-    await this.cryptoApi.cacheKeyPair(jwk)
-    await this.keyStorage.storeKeyPair(`${dataOwnerId}.${publicKey.slice(-32)}`, jwk)
-
     const dataOwner = await this.cryptoApi.getDataOwner(dataOwnerId).catch((e) => {
       throw this.errorHandler.createErrorFromAny(e)
     })
+
     if (dataOwner == null) {
       throw this.errorHandler.createErrorWithMessage(
         `The current user is not a data owner. You must be either a patient, a device or a healthcare professional to call this method.`
       )
     }
 
+    const dataOwnerPublicKeys = this._getDataOwnerPublicKeys(dataOwner.dataOwner)
+
+    const fetchCurrentUserKeyPairOnDevice = async () => {
+      if (dataOwnerPublicKeys.length <= 0) {
+        return undefined
+      }
+
+      const keyPairs = (await Promise.all(dataOwnerPublicKeys.map((pub) => this.keyStorage.getKeypair(`${dataOwnerId}.${pub.slice(-32)}`)))).filter(
+        (kp) => kp !== undefined
+      ) as { publicKey: JsonWebKey; privateKey: JsonWebKey }[]
+
+      await Promise.all(keyPairs.map((kp) => this.cryptoApi.cacheKeyPair(kp)))
+      const currentKeyPair = keyPairs.find((kp) => jwk2spki(kp.publicKey) === dataOwner.dataOwner.publicKey!)
+
+      return currentKeyPair !== undefined
+        ? { privateKey: jwk2pkcs8(currentKeyPair.privateKey), publicKey: jwk2spki(currentKeyPair.publicKey) }
+        : undefined
+    }
+
+    const currentUserKeyPairOnDevice = await fetchCurrentUserKeyPairOnDevice()
+
+    if (currentUserKeyPairOnDevice !== undefined) {
+      if (
+        keyPair !== undefined &&
+        (currentUserKeyPairOnDevice.publicKey !== keyPair.publicKey || keyPair.privateKey !== currentUserKeyPairOnDevice.privateKey)
+      ) {
+        throw this.errorHandler.createErrorWithMessage(
+          `The current user already has a keypair on the device. You must either provide the same keypair or must not provide any keypair.`
+        )
+      }
+      return currentUserKeyPairOnDevice
+    }
+
+    const { publicKey, privateKey } = keyPair ?? (await this._generateKeyPair())
```

Comment:
If you found some keys and cached them into the crypto, you don't even need to go through it, storeKeypair is useless and cacheKeypair is duplicated. 
Actually, if you found your key(s) into the keyStorage, you're done once you cached it. Simplify your code